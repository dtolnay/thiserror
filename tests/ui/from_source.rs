use std::io;
use thiserror::Error;

#[derive(Error, Debug)]
struct BracedWithSourceDuplicate {
    source: io::Error,
    // this should cause compilation error
    // becuase there's two `source`
    #[source]
    cause: io::Error,
}

#[derive(Error, Debug)]
struct BracedWithFromDuplicate1 {
    #[from]
    source: io::Error,
    // this should cause compilation error
    // because there's two `from`
    #[from]
    cause: io::Error,
}

#[derive(Error, Debug)]
struct BracedWithFromDuplicateSource {
    source: io::Error,
    // this should cause compilation error
    // because from implies source
    //
    // TODO should this error actually be that there's a from tha't not on source?
    #[from]
    cause: io::Error,
}

#[derive(Error, Debug)]
struct BracedWithFromExtraField1 {
    #[from]
    source: io::Error,
    // this should cause compilation error
    // because fields other than `from` are not allowed (except backtrace in future)
    msg: String,
}

#[derive(Error, Debug)]
enum EnumDuplicateFromType1 {
    BracedOne {
        #[from]
        cause: io::Error,
    },
    Tuple(#[from] io::Error),
    Unit,
}

#[derive(Error, Debug)]
enum EnumDuplicateFromType2 {
    BracedOne {
        #[from]
        cause: io::Error,
    },
    Tuple(#[from] std::io::Error),
    Unit,
}

fn main() {}
