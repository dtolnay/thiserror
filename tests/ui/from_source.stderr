error: Only one `source` field allowed per struct or struct variant (remember that `#[from]` implies `source`)
  --> $DIR/from_source.rs:10:5
   |
10 |     cause: io::Error,
   |     ^^^^^

error: Only one `#[from]` field allowed per struct or struct variant
  --> $DIR/from_source.rs:20:5
   |
20 |     cause: io::Error,
   |     ^^^^^

error: When deriving `From`, non-`#[from]` fields are not allowed
  --> $DIR/from_source.rs:25:5
   |
25 |     source: io::Error,
   |     ^^^^^^

error: When deriving `From`, non-`#[from]` fields are not allowed
  --> $DIR/from_source.rs:40:5
   |
40 |     msg: String,
   |     ^^^

error: Cannot derive `From` on enum when two variants have same type for source error
  --> $DIR/from_source.rs:49:19
   |
49 |     Tuple(#[from] io::Error),
   |                   ^^^^^^^^^

error: Cannot derive `From` on enum when two variants have same type for source error
  --> $DIR/from_source.rs:59:19
   |
59 |     Tuple(#[from] std::io::Error),
   |                   ^^^^^^^^^^^^^^
